(QENG-4823) AWS EC2 Reaper

tvpartytonight opened the pull request:
This is the initial PR for the AWS EC2 Reaper.

Commit:
(QENG-4823) AWS EC2 Reaper

This is the initial commit for the AWS EC2 Reaper.

Commit:
(QENG-4823) Add safety for destructive methods

Previous to this commit, the AWS EC2 reaper did not have an easy way to
turn on or off the destructive capabilities; this PR adds a global
switch that can be toggled by an environment variable that controls
whether or not the reaper is running with destructive capabilities.

ferglor commented on lambdas/ec2/README.md:
```diff
@@ -0,0 +1,42 @@
+## Rules and Usage
+
+1. The Schema Enforcer ensures that newly created EC2 instances have a valid `termination_date` tag set. If a valid `termination_date` is not set within 4 minutes, the Schema Enforcer will terminate the instance and raise an exception. The `termination_date` must be a valid IS0 8601 value with a UTC offset defined.
+    * Instead of setting the `termination_date` directly, you can set a `lifetime` tag to specify how long the instance should live. A valid value is a string of of an integer value with a 1 letter unit of w(weeks), d(days), h(hours). For example, `1w` is 1 week, `2d` is 2 days, and `3h` is 3 hours. The Schema Enforcer will caculcate a future date based upon the `lifetime` tag and set a new `term ...
```
A valid value is a string **of of** an integer...

nwolfe replied:
s/caculcate/calculate

nwolfe replied:
> the Schema Enforcer will terminate the instance and raise an exception

What does "raise an exception" mean for a user of AWS?

Overall, is this "Rules and Usage" section targeted at a user or a developer?  Seems like it's trying to be for the user, and the mention of exceptions is throwing me off because I don't know what behavior the user would experience because of a thrown exception.

tvpartytonight replied:
Good point @nwolfe, I'll make a cleaner separation between the AWS Reaper Rules documentation and the AWS Reaper admin documentation.

nwolfe replied:
Wait....I'm confused which tag I should be setting as a user. 

In our IRL talks I thought I would be setting the `lifetime` tag since that's easier to think about (e.g. "I need an EC2 instance for 24 hours").

But almost all this documentation is talking about the `termination_date`.

tvpartytonight replied:
There's definitely more mention of the termination_date, so I'll fix up the User section to encourage them to use the lifetime tag.

ferglor commented on lambdas/ec2/README.md:
```diff
@@ -0,0 +1,42 @@
+
+The Terminator should run on a Cloudwatch schedule, configurable in the AWS EC2 GUI. 
+
+Both Lambdas do not actually delete instances unless they are running in an environment where the `LIVE_MODE` environment variable is defined. To turn these Lambdas on and allow them to actually terminate instances, they must run in an environment where `LIVE_MODE` is defined. Please note that defining `LIVE_MODE` to `False` does not turn off the Lambdas; the environment variable must be removed  ...
```
I would say "defining `LIVE_MODE` **as** `False`" or "**setting** `LIVE_MODE` to `False`"

nwolfe replied:
So the actual value of `LIVE_MODE` doesn't matter?  Just the existence of it?  

That seems dangerous! 

And it could lead to 💵  down the 🚽  right?

Unless I'm misunderstanding something we should rework this variable behavior & name.

tvpartytonight replied:
I'm not sure what you mean by 💰 down the 🚽 , but yeah, this is a little bit dangerous. 🗡

nwolfe replied:
I just mean that this flag basically controls whether the Reaper does anything or not, and the fact that it's a boolean value where the value actually doesn't matter, could lead to the admin thinking that they can turn it on/off by setting it to `True`/`False` when in fact that won't actually change the behavior.

ferglor commented on lambdas/ec2/schema_enforcer.py:
```diff
@@ -0,0 +1,198 @@
+    if exception is None:
+        raise Exception(message)
+    else:
+        Warning(message)
```
How does this differ from warnings.warn()?

tvpartytonight replied:
this should be warn. good catch.

nwolfe commented on lambdas/ec2/README.md:
```diff
@@ -0,0 +1,42 @@
+The Schema Enforcer is a lambda that is designed to be triggered when an EC2 instance enters the 'pending' state. This lambda waits for an EC2 instance to have a valid `termination_date` tag associated with it. This Lambda also listens for a `lifetime` tag; if found, it calculates a new future date and adds that date as the `termination_date` for the instance. The Schema Enforcer terminates insta ...
+
+### Terminator
+The Terminator is a simple lambda that looks for a `termination_date` tag on an instance and terminates it if it is past its `termination date`. If the `termination_date` is missing or malformed, the script reports those instances in an exception. This lambda is designed to be run periodically; depending on your needs, every 15 minutes should be more than sufficient. The Terminator does not ensur ...
```
\`termination date\` -> \`termination_date\`?

nwolfe commented on lambdas/ec2/README.md:
```diff
@@ -0,0 +1,42 @@
+    * Instead of setting the `termination_date` directly, you can set a `lifetime` tag to specify how long the instance should live. A valid value is a string of of an integer value with a 1 letter unit of w(weeks), d(days), h(hours). For example, `1w` is 1 week, `2d` is 2 days, and `3h` is 3 hours. The Schema Enforcer will caculcate a future date based upon the `lifetime` tag and set a new `term ...
+2. The Terminator runs periodically to ensure that all EC2 instances are terminated if they are past their 'termination_date'. If the lifetime for an instance needs to be prolonged, the `termination_date` should be updated directly.
+
+## Installation 
```
Would a user of AWS need to read this section, or is this section just for the admin who is installing the Reaper on the AWS instance?

It seems like maybe we have 2 separate audiences to speak to: the AWS user and the AWS admin.

If that's true, it might be nice to separate the docs for each a bit more.

nwolfe commented on lambdas/ec2/README.md:
```diff
@@ -0,0 +1,42 @@
+2. The Terminator runs periodically to ensure that all EC2 instances are terminated if they are past their 'termination_date'. If the lifetime for an instance needs to be prolonged, the `termination_date` should be updated directly.
+
+## Installation 
+Currently, there is no build script yet for the Reaper; you will need to copy the `schema_enforcer.py` and `terminator.py` files to Lambdas in AWS. You will also need to ensure that there is a role for the lambda's with sufficient privilege and access to read events and delete instances. Once those are in place, you will need to add a rule in Cloudwatch for the Enforcer with this event pattern:
```
> you will need to add a rule in Cloudwatch

Is this common knowledge in AWS land, or is a link worthwhile here?  I'm totally new to AWS so figured I'd ask just in case.

nwolfe commented on lambdas/ec2/README.md:
```diff
@@ -0,0 +1,42 @@
+
+1. The Schema Enforcer ensures that newly created EC2 instances have a valid `termination_date` tag set. If a valid `termination_date` is not set within 4 minutes, the Schema Enforcer will terminate the instance and raise an exception. The `termination_date` must be a valid IS0 8601 value with a UTC offset defined.
+    * Instead of setting the `termination_date` directly, you can set a `lifetime` tag to specify how long the instance should live. A valid value is a string of of an integer value with a 1 letter unit of w(weeks), d(days), h(hours). For example, `1w` is 1 week, `2d` is 2 days, and `3h` is 3 hours. The Schema Enforcer will caculcate a future date based upon the `lifetime` tag and set a new `term ...
+2. The Terminator runs periodically to ensure that all EC2 instances are terminated if they are past their 'termination_date'. If the lifetime for an instance needs to be prolonged, the `termination_date` should be updated directly.
```
Did you mean for this to be syntax-highlighted all the other occurrences of this?

'termination_date' -> \`termination_date\`

nwolfe commented:
@tvpartytonight I'll wait to review the Python once you've made the changes you were discussing IRL.

Nice work though!

kevpl commented on lambdas/ec2/schema_enforcer.py:
```diff
@@ -0,0 +1,198 @@
+from __future__ import print_function
+
+import json
```
do you actually use the `json` module in any of this code? Is there a reason to be including modules you're not using?

tvpartytonight replied:
yeah, that's leftover and can be removed.

Commit:
QENG-4823) Combine scripts and improve docs

This commit combines the scripts into a single resource to limit
duplication; this means that the default lambda_handler definition must
change to `enforce` and `terminate`. Also included in this PR is a
cleanup of the documentation, where there is a cleaner separation for
docs for AWS users and docs for people implementing the Reaper.

tvpartytonight commented:
OK, I believe I have added changes that address all the comments so far; ready for further review!

mchllweeks commented on lambdas/ec2/README.md:
```diff
@@ -1,15 +1,47 @@
+value with a 1 letter unit of w(weeks), d(days), h(hours). For example, `1w` is 
+1 week, `2d` is 2 days, and `3h` is 3 hours. The Schema Enforcer will calculcate 
+a future date based upon the `lifetime` tag and set a new `termination_date` tag 
+on the instance that.
```
Did you mean "set a new 'termination_date' tag on that instance"?

mchllweeks commented on lambdas/ec2/README.md:
```diff
@@ -1,15 +1,47 @@
+2. If there is an error determining the future termination date, the instance is 
+terminated. If 4 minutes elapse and no future termination date has been 
+determined, the instance is terminated.
+2. The Terminator runs periodically to ensure that all EC2 instances are
```
should this be "3."?

mchllweeks commented on lambdas/ec2/README.md:
```diff
@@ -27,16 +59,39 @@ Currently, there is no build script yet for the Reaper; you will need to copy th
+
+To turn these AWS Lambdas on and allow them to actually terminate instances, 
+they must run in an environment where `LIVE_MODE` is defined as true. Valid 
+`LIVE_MODE`values are case-insensitive 'true'or 'false' strings. If `LIVE_MODE` 
```
space missing after 'true'

mchllweeks commented on lambdas/ec2/README.md:
```diff
@@ -27,16 +59,39 @@ Currently, there is no build script yet for the Reaper; you will need to copy th
+exception raised. If the Schema Enforcer encounters a handled error during 
+executuion, it terminates the instance and raises an error. Unhandled errors are
+raised, but the Schema Enforcer does not terminate the instance in these cases.
+The Schema Enforcerdoes not terminate instances after the schema has been 
```
space missing after 'Enforcer'

mchllweeks commented on lambdas/ec2/reaper.py:
```diff
@@ -11,10 +10,26 @@
+def determine_live_mode():
+    """
+    Returns True if LIVE_MODE is set to true in the shell environment, False if
+    set to false or if not set at all. Raises error if LIVE_MODE is set any
```
'set as any'

tvpartytonight replied:
I'm not sure what you re suggesting here...

mchllweeks replied:
I wasn't sure if 'Raises error if LIVE-MODE is set any other string' had a typo and you meant 'is set as any other string' or 'is set to any other string'.

tvpartytonight replied:
ahh, I was looking in the wrong section of the string. 👓

mchllweeks commented on lambdas/ec2/reaper.py:
```diff
@@ -169,7 +184,9 @@ def calculate_lifetime_delta(lifetime_value):
-def lambda_handler(event, context):
+# This is the function that the schema_enforcer lambda should run when an instance hits
+# the pending state.
+def enforce(event, context):
```
doc string?

mchllweeks commented on lambdas/ec2/reaper.py:
```diff
@@ -196,3 +213,42 @@ def lambda_handler(event, context):
+
+# This is the function that a terminator lambda should call periodically to delete instances past their
+# termination_date.
+def terminate(event, context):
```
doc string?

nwolfe commented on lambdas/ec2/reaper.py:
```diff
@@ -0,0 +1,268 @@
+                    }
+                ]
+            )
+    else:
```
WTF is a while-else!?!

tvpartytonight replied:
https://docs.python.org/2/reference/compound_stmts.html#while

> A break statement executed in the first suite terminates the loop without executing the else clause’s suite. A continue statement executed in the first suite skips the rest of the suite and goes back to testing the expression.

Commit:
(QENG-4823) Remove else from while condition in wait_for_tags

This change removes the unusual else behavior from the wait_for_tags method and
changes the while condition to return instead of break when a termination_date
is found.
